Documentation/ryujit-tutorial.md

```diff
@@ -0,0 +1,728 @@
+  - Transforms while loops to “do while”
+  - Performs loop cloning and unrolling
+  - Loops may be invalidated, but must be marked as such. 
+
+### LclVar Sorting and Tree Ordering
+- lvaMarkLocalVars()
+  - Set reference counts, sort and select for tracking
+- Execution order within a statement is set by fgSetBlockOrder()
+  - Canonical postorder traversal of the nodes
+  - Execution order is made concrete by adding gtNext/gtPrev links
+  - Must also correspond with canonical order: “Left to right” except for binary operators with GTF_REVERSE_OPS flag
+  - If op1 has no side-effects, a simple cost analysis is done to determine when to reverse the order of binary operands
+  - A GTF_REVERSE_OPS flag on an assignment indicates that the rhs (op2) should be evaluated before the target address (if any) on the lhs (op1) is evaluated.
+
+### Tree Order
+- Initially, expression nodes are linked only via parent-child links (uni-directional)
+- The consuming node (parent) has pointers to the nodes that produce its input operands (children)
+- Execution order is implicit
+  - Operands are evaluated in order
+  - Unless the GTF_REVERSE_OPS flag is set
+
+### RyuJIT IR
+- Before ordering
+** Pre-ordering IR diagram 
+
+#### Notes
+As a reminder, this is a depiction of the graph before ordering, in what we call “Tree Order” mode
+
+### Linear Order
+- After Ordering, IR expression nodes have both parent-child links as well as execution order links (gtPrev and gtNext)
+- The execution order no longer represents a canonical traversal of the parent ->  child links
+- But the relative execution order of the children must be maintained
+- Statements are replaced with GT_IL_OFFSET nodes to allow mapping back to the incoming IL
+- 
+### RyuJIT IR
+After ordering
+** Post-ordering IR diagram 
+
+#### Notes
+This is the same diagram as before, but with additional links to indicate execution order. The statements have a link to the first node in the list, and each node has a link to its previous and next node in the linear order.
+
+## RyuJIT Front-End
+
+### Loop Invariant Code Hoisting
+- Identifies simple loops
+- Traverses loops outer-to-inner, hoisting expressions that
+  - Have not been marked GTF_DONT_CSE
+  - Have no side-effects
+  - Don’t raise exceptions, OR occur before any side-effects
+  - Have a valid value number and are invariant:
+    - Constants
+    - LclVars defined outside the loop
+    - Nodes whose children are invariant
+
+### Copy Propagation
+- Walks each block in the graph in dominator-first order
+  - When it encounters a variable with the same VN as a previously live variable, it is replaced with that variable
+
+### CSE & Assertion Propagation
+- Common Subexpression Elimination
+  - Redundant expressions are identified by value number
+  - If deemed profitable, they are evaluated to a new “temp” lclVar and then reused
+- Assertion Propagation
+  - Propagate properties such as constant values and non-nullness
```

Comment:
Perhaps it should be mentioned by assertion propagation does eliminate some range checks as well